### DISCUSSION TITLE
changing transforms outside of animate

### DISCUSSION THREAD

**Discussion by akbr** [motiondivision/motionone#16] [Ideas]
`animate` uses its own css variables with `style.transform` to enable individual transforms.

But this took me by surprise:

```js
el.style.transform = "translateX(100px)";
/** other styles **/

// ... later ...

animate(el, {rotate: 45});
// translateX is lost

// from now on, I have to remember that motion "owns" style.transform on this element (through its --motion-* css vars)
// if I try to change the transform with any other method, I'll mess up the styles
```

Because individual transforms are so useful, it'd be nice to opt in even when I don't want an animation. For example:

```js
import {style, animate} from "motion";

style(el, {x:100 /** other styles **/});
// ... later ...
animate(el, {rotate: 45});
```

I don't know if that's API heresy for a library called motion, though.... 😂

**Comment by okikio** [comment]
That seems like a neat idea

**Comment by mattgperry** [comment]
No it’s a great idea, I agree, I don’t know if I’ve done it yet but was internally going to have a style.get/set that handles this so I can export that

**Comment by mattgperry** [comment]
Also it’s not so crazy, I believe GSAP has similar

**Reply by mikehwagz** [reply]
Yes! `.set` is this is one of the only remaining GSAP things I still find myself reaching for when using this lib. I really like the ability to use `.set` within GSAP timelines as well.

**Comment by subtleGradient** [comment]
Tailwind CSS already uses css vars for all the separate transforms. It'd be awesome if Motion One could seamlessly integrate somehow.

https://tailwindcss.com/docs/transform

**Reply by subtleGradient** [reply]
Maybe integrating with Tailwind is as easy as using motion to animate their transform css vars?
Would this work?

```js
animate(el, {
    '--tw-translate-x': 0,
    '--tw-translate-y': 0,
    '--tw-rotate': 0,
    '--tw-skew-x': 0,
    '--tw-skew-y': 0,
    '--tw-scale-x': 1,
    '--tw-scale-y': 1,
})
```

**Reply by okikio** [reply]
That could cause unexpected bugs.

**Reply by subtleGradient** [reply]
I guess alternatively I could configure Tailwind to use the motion transform variable names

**Reply by okikio** [reply]
Yeah, I think that would be better

**Comment by mattgperry** [comment]
I haven't documented this yet but you can do `style.set(element, name, value)` and `style.get(element, name)`

**Comment by akbr** [comment]
This is trickier than I first realized. 

At the end of the day, I'd love to use Motion One's `keyframes` format to apply styles synchronously _and_ with WAAPI _interchangeably_. 

Conceptually, I want to be able to write `animate(el, {x: 100}, {duration: "DO IT NOW!"})` and set the style `"transform: translateX(var(--motion-translateX)); --motion-translateX:100px;"` synchronously.

`style.set(el, "x", 100)` sets `--motion-translateX:100`, but doesn't hook into conventions like unit conversion and setting the base transform string.

Is there an elegant way to expose a function that takes in `keyframes`, and returns the final styles that would be committed at the end of an animation?

**Reply by mattgperry** [reply]
Ah yeah good point, I can probably figure it out that style.set does this

**Reply by akbr** [reply]
after some more thought, I've decided that (1) this is not trivial, and (2) `style.set` may not be the right solution 😂

what I'm looking for, as @mikehwagz [mentioned](https://github.com/motiondivision/motionone/discussions/16#discussioncomment-1903558) above, is an equivalent to [GSAP's .set method](https://greensock.com/docs/v3/GSAP/Tween/set()), which is "essentially a zero-duration [synchronous] to() tween."

as a DX matter, I want to apply whole `keyframe` blocks synchronously. (even if they contain an array of values for a property.)

to be maintainable, it'd seem like that API would have to follow the `animate/animate-style.ts` code path, or a modularized piece of it.

it'd be easy to slap some `./utils/transforms` logic onto `style.set`, but I worry that wouldn't be a very robust  or maintainable solution. it also seems a little odd for a function called `style.set` to be registering CSS properties, reading individual transforms from an element, etc.
